src/llmq/context.cpp

```diff
@@ -20,108 +20,81 @@
-    bls_worker = std::make_shared<CBLSWorker>();
-
-    dkg_debugman = std::make_unique<llmq::CDKGDebugManager>();
-    llmq::quorumBlockProcessor = std::make_unique<llmq::CQuorumBlockProcessor>(evo_db, connman, peerman);
-    qdkgsman = std::make_unique<llmq::CDKGSessionManager>(connman, *bls_worker, *dkg_debugman, *llmq::quorumBlockProcessor, sporkman, peerman, unit_tests, wipe);
-    llmq::quorumManager = std::make_unique<llmq::CQuorumManager>(evo_db, connman, *bls_worker, *llmq::quorumBlockProcessor, *qdkgsman, ::masternodeSync, peerman);
-    sigman = std::make_unique<llmq::CSigningManager>(connman, *llmq::quorumManager, peerman, unit_tests, wipe);
-    shareman = std::make_unique<llmq::CSigSharesManager>(connman, *llmq::quorumManager, *sigman, peerman);
-    llmq::chainLocksHandler = std::make_unique<llmq::CChainLocksHandler>(mempool, connman, sporkman, *sigman, *shareman, *llmq::quorumManager, *::masternodeSync, peerman);
-    llmq::quorumInstantSendManager = std::make_unique<llmq::CInstantSendManager>(mempool, connman, sporkman, *llmq::quorumManager, *sigman, *shareman, *llmq::chainLocksHandler, *::masternodeSync, peerman, unit_tests, wipe);
-
+                         const std::unique_ptr<PeerManager>& peerman, bool unit_tests, bool wipe) :
+    bls_worker{std::make_shared<CBLSWorker>()},
+    dkg_debugman{std::make_unique<llmq::CDKGDebugManager>()},
+    quorum_block_processor{[&]() -> llmq::CQuorumBlockProcessor* const {
+        llmq::quorumBlockProcessor = std::make_unique<llmq::CQuorumBlockProcessor>(evo_db, connman, peerman);
+        return llmq::quorumBlockProcessor.get();
+    }()},
+    qdkgsman{std::make_unique<llmq::CDKGSessionManager>(connman, *bls_worker, *dkg_debugman, *quorum_block_processor, sporkman, peerman, unit_tests, wipe)},
+    qman{[&]() -> llmq::CQuorumManager* const {
+        llmq::quorumManager = std::make_unique<llmq::CQuorumManager>(evo_db, connman, *bls_worker, *quorum_block_processor, *qdkgsman, ::masternodeSync, peerman);
+        return llmq::quorumManager.get();
+    }()},
+    sigman{std::make_unique<llmq::CSigningManager>(connman, *llmq::quorumManager, peerman, unit_tests, wipe)},
+    shareman{std::make_unique<llmq::CSigSharesManager>(connman, *llmq::quorumManager, *sigman, peerman)},
+    clhandler{[&]() -> llmq::CChainLocksHandler* const {
+        llmq::chainLocksHandler = std::make_unique<llmq::CChainLocksHandler>(mempool, connman, sporkman, *sigman, *shareman, *llmq::quorumManager, *::masternodeSync, peerman);
+        return llmq::chainLocksHandler.get();
+    }()},
+    isman{[&]() -> llmq::CInstantSendManager* const {
+        llmq::quorumInstantSendManager = std::make_unique<llmq::CInstantSendManager>(mempool, connman, sporkman, *llmq::quorumManager, *sigman, *shareman, *llmq::chainLocksHandler, *::masternodeSync, peerman, unit_tests, wipe);
+        return llmq::quorumInstantSendManager.get();
+    }()}
```

Comment:
I want to note that these only work as we're assigning the global here; at first look it makes it seem that the unique_ptrs should be destroyed at the end of the lambda, but they're not since they're assigned to the global.

Reply:
I'm also concerned that this will or could result in double initialization of these globals...

Reply:
I refactored this part to show intentions of initialize both class-member and global object more clear.

Also, there's added assert to be sure that global variables are non-double initialized anyhow.